crux/src/Crux/Log.hs

```diff
@@ -25,23 +38,154 @@ module Crux.Log (
+
+cruxWarn :: T.Text -> SayWhat
+cruxWarn = SayWhat Warn cruxLogTag
+
+
+cruxMessageToSayWhat :: SayWhatHandler ext => CruxLogMessage ext -> IO SayWhat
+
+cruxMessageToSayWhat AttemptingProvingVCs =
+  pure (cruxSimply "Attempting to prove verification conditions.")
+
+cruxMessageToSayWhat (Checking files) =
+  pure (cruxNoisily ("Checking " <> T.intercalate ", " (T.pack <$> files)))
+
+cruxMessageToSayWhat (ExtensionLogMessage ext) = ?extToSayWhat ext
+
+cruxMessageToSayWhat FoundCounterExample =
+  pure (cruxOK "Counterexample found, skipping remaining goals")
+
+cruxMessageToSayWhat (Goals _) = pure SayNothing -- message for IDE only
+
+cruxMessageToSayWhat (Help (LogDoc doc)) = do
+  outWidth <- maybe 80 (\(Window _ w) -> w) <$> size
+  let opts = LayoutOptions $ AvailablePerLine outWidth 0.98
+  return (cruxOK (renderStrict (layoutPretty opts doc)))
+
+cruxMessageToSayWhat SimulationComplete =
+  pure (cruxNoisily "Simulation complete.")
+
+cruxMessageToSayWhat SimulationTimedOut =
+  pure (cruxWarn
+    "Simulation timed out! Program might not be fully verified!"
+  )
+
+cruxMessageToSayWhat (StartedGoal goalNumber) =
+  pure (cruxOK (T.pack ("Started goal " ++ show goalNumber)))
+
+cruxMessageToSayWhat (TotalPathsExplored i) =
+  pure (cruxSimply ("Total paths explored: " <> T.pack (show i)))
+
+cruxMessageToSayWhat (Version nm ver) = do
+  let text = unwords
+        [ "version: " <> version <> ","
+        , T.unpack nm
+        , "version: " <> (showVersion ver)
+        ]
+  return (cruxOK (T.pack text))
+
+cruxMessageToSayWhat WarnDisablingBranchCoverageRequiresPathSatisfiability =
+  pure (cruxWarn
+    "Branch coverage requires enabling path satisfiability checking.  Coverage measurement is disabled!"
+  )
+
+cruxMessageToSayWhat WarnDisablingProfilingIncompatibleWithPathSplitting =
+  pure (cruxWarn
+    "Path splitting strategies are incompatible with Crucible profiling. Profiling is disabled!"
+  )
+
+cruxMessageToSayWhat (WarnPathsUnexplored numberOfPaths) =
+  pure (cruxWarn (T.unwords
+    [ T.pack $ show numberOfPaths
+    , "paths remaining not explored: program might not be fully verified"
+    ]))
+
+cruxMessageToSayWhat WarnSkippingUnsatCoresBecauseMCSatEnabled =
+  pure (cruxWarn "Warning: skipping unsat cores because MC-SAT is enabled.")
+
+cruxMessageToSayWhat (WarnUnsupportedTimeoutFor backend) =
+  pure (
+    cruxWarn
+    (T.pack ("Goal timeout requested but not supported by " <> backend))
+  )
+
+
+-- | Main function used to log/output a general text message of some kind
+say :: Logs => CruxLogMessage () -> IO ()
+say =
+  let ?extToSayWhat = \ () -> pure SayNothing in
+  LJ.writeLog (?outputConfig ^. logWhat) <=< cruxMessageToSayWhat
+
 -- | Main function used to log/output a general text message of some kind
-say :: Logs => SayLevel -> Text -> Text -> IO ()
-say lvl frm = LJ.writeLog (?outputConfig ^. logWhat) . SayWhat lvl frm
+sayExt :: Logs => SayWhatHandler ext => ext -> IO ()
```

Comment:
Could we use something other than "ext" to represent the front-end?  Using `ext` conflates with the crucible `ext` and they are fairly different.  The closest analogy would probably be the crucible `personality` parameter, but I have a hard time coming up with a function name to use here (and related) that is a short form for personality.   You could just use something like `says` or `sayCruxMsg`... or anything else that seems reasonable.

Reply:
Yeah we can definitely change this. I don't love `says`, because it reminds me of `get`/`gets`, `put`/`puts`, but does not have the same semantics.

Will think of something.

Reply:
New commit shows a proposed change:
https://github.com/GaloisInc/crucible/blob/3c333baf99c624b0810d5ee72294f22c4952cbd3/crux/src/Crux/Log.hs#L99-L105

`say` is now polymorphic over the message data type, but has a constraint for an implicit parameter turning that message into a `SayWhat`.